spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/code/InMemoryAuthorizationCodeServices.java

```diff
@@ -11,18 +12,52 @@
  * @author Dave Syer
  */
 public class InMemoryAuthorizationCodeServices extends RandomValueAuthorizationCodeServices {
+	/**
+	 *  From RFC6749: A maximum authorization code lifetime of 10 minutes is RECOMMENDED
+	 *
+	 *  @see <a href="https://tools.ietf.org/html/rfc6749#section-4.1.2">RFC6749 4.1.2.  Authorization Response</a>
+	 */
+	private static final int DEFAULT_CODE_LIFETIME_SECONDS = 10*60;
 
-	protected final ConcurrentHashMap<String, OAuth2Authentication> authorizationCodeStore = new ConcurrentHashMap<String, OAuth2Authentication>();
+	private int codeLiftetimeSeconds = DEFAULT_CODE_LIFETIME_SECONDS;
+
+	protected final ConcurrentHashMap<String, ExpiringOAuth2Authentication> authorizationCodeStore = new ConcurrentHashMap<String, ExpiringOAuth2Authentication>();
 
 	@Override
 	protected void store(String code, OAuth2Authentication authentication) {
-		this.authorizationCodeStore.put(code, authentication);
+		this.authorizationCodeStore.put(code, new ExpiringOAuth2Authentication(System.currentTimeMillis(),
+			authentication));
 	}
 
 	@Override
 	public OAuth2Authentication remove(String code) {
-		OAuth2Authentication auth = this.authorizationCodeStore.remove(code);
-		return auth;
+		removeExpired();
+		ExpiringOAuth2Authentication expiringOAuth2Authentication = this.authorizationCodeStore.remove(code);
+		return expiringOAuth2Authentication == null ? null : expiringOAuth2Authentication.authentication;
+	}
+
+	private void removeExpired() {
```

Comment:
I don't feel the expiring logic should be in this implementation. Instead, it should be in an external component that can be used for any `AuthorizationCodeServices` implementation. There is similar logic in `JdbcAuthorizationCodeServices` (in this PR) and `RedisAuthorizationCodeServices` would have to contain similar logic as well. As you can see, there is duplication of efforts across `AuthorizationCodeServices` implementations.